### DISCUSSION TITLE
CMS media files: security benefits of file storage services vs. do-it-yourself file servers

### DISCUSSION THREAD

**Discussion by thibaudcolas** [wagtail/wagtail#10964] [General]
_Copying a discussion thread started in Slack for future reference – and possibly wider input_

Hi #security, I could use your help with site hosting documentation I’m working on, specifically a section about CMS-uploaded images and documents (Django covers this in their [User-uploaded content](https://docs.djangoproject.com/en/4.2/topics/security/#user-uploaded-content) docs).

The specific security aspect I’m trying to figure out is how much it’s "worth the trouble" to store and serve those files from a dedicated file storage service (like S3), as opposed to a VPS-style setup where the files are stored on a file system and served by say Apache or Nginx from the site’s main domain.

From my understanding, by using a separate files storage service you get the following _security_ benefits:

1. The big one – Files are served from a separate domain, so the same-origin policy will prevent a range of exploit types that aren’t really possible to fully prevent otherwise (could use antivirus-type software to help but it’s an arms race)
2. There are no security risks of your file server being misconfigured in a way that allows uploaded files to execute.
3. Since your files are already well separate from the rest of your site, it’s simpler to scale horizontally or host in multiple locations to help with your site’s availability / resilience.
4. Feels flimsy but I guess it counts – it’s one less way to DOS your site. Though someone could DOS the file storage provider, only your site’s images would be missing.

Is that the right order? Are there any other clear security benefits?

Based on this list, there are two counterpoints which I don’t know what to think about:

1. The first two points only apply if you don’t trust the file uploaders. Which is a good default position to be in, but not super relevant if you set up a CMS which only you use for a personal site (or even say three people in a company).
2. The last two points only apply if the site is business-critical enough to require high availability. So again probably not super relevant for personal / small-company sites.

So for a personal or otherwise "small" site, with fully trusted CMS users and high availability not a critical concern – would you still use a separate file storage service?

**Comment by thibaudcolas** [comment]
Replies so far:

> @RealOrangeOne: My personal website serves its static / media files with nginx on the same server as the application. If it's served with nginx rather than Django itself, even a tiny server will be able to handle a lot of concurrent file requests without issue.
> 
> A sensible nginx config to serve the files is also nice and simple: https://github.com/RealOrangeOne/website/blob/master/etc/nginx.conf#L29-L38

> @spapas: Please notice that if you have lots of files back ups etc would be problematic in the self served case (in S3 it isn't relevant). same if you want to change servers

> @thibaudcolas: Hmm this isn’t something I know much about! I assume if you know enough to set up your own file storage, you probably know what to do about backups, no?

> @RealOrangeOne: Theoretically, yes. But then, if you have enough files to warrant that scale, you probably should be outsourcing it

**Comment by thibaudcolas** [comment]
Separate thread:

> @allcaps: A downside of S3 compatible storage (external storage) is the additional complexity. The chance of misconfiguration is bigger. So all the pros can be pitfalls too. It certainly requires more thought and consideration over serving the files from your own server.
> Maybe there is also something to say about private files and content policies?

> @thibaudcolas: I’d agree there’s a chance of misconfiguration but that doesn’t feel like a security downside overall unless there are specific types of misconfiguration that would affect site integrity / availability / privacy?
> Private files is an interesting one – is that capability a security benefit of file storage/serving services, or serving from one’s own server, or neither?
> For content policies – I suppose it’s a configuration complexity factor too but there’s nothing that’s inherently better either way?

> Replying to: Private files is an interesting one – is that capability a security benefit of file storage/serving services, or serving from one’s own server, or neither?

> @allcaps: Neither, if you know what you are doing.
> - Simplest: files over the Django process. This occupies your Django process with potential long running download, and that is a risk. It requires your uploads to go in a directory that is private (not served by your webserver).
> - Better: x-sendfile to handle permission checks by Django process and offload the serving of the file to your webserver. NGINX or Apache does the heavy lifting. Same as before, the files must be private.
> - With a S3, you need Access Control Lists (ACLs) or Bucket policies. The identity and access management is with Django process, it will redirect to a S3 Pre-Signed URL. These URLs have a limited validity period and are signed with AWS credentials, allowing temporary access to the designated file.
>
> It is true that with any of these, misconfiguring can expose files. But, if we give advice, I'd say to start at the shallow end of the pool.

> @thibaudcolas: Thanks, sendfile is one of those things I know nothing about


> @spapas: please be careful; "Simplest" is not simplest, it is plain wrong because it will always break your site
never ever serve files over a django  python process


> @allcaps: This explains it https://www.nginx.com/resources/wiki/start/topics/examples/xsendfile/
The trick is to store your media outside your public media directory.

```python
def get_private_storage(*args, **kwargs):
    return FileSystemStorage(location=settings.PRIVATE_ROOT)


class CustomImage(AbstractImage):
    # Override file, use private storage.
    file = models.ImageField(
        verbose_name=_("file"),
        upload_to="original_images/%Y/%m/%d/",
        storage=get_private_storage,
        width_field="width",
        height_field="height",
        max_length=255,
    )
````

> @spapas: the really simplest (and secure enoguh) method is to serve them using something like this https://pypi.org/project/django-random-filestorage/ ; which leaves the files in the open but the filenames can't be guessed

> @thibaudcolas: Do you have a good example of the real-world use case for private images? I’ve heard of a few in the past but forgot what they were

> @allcaps: Never say never :wink:. A Wagtail export download is serving a file through Django, and that is fine.
So files over Django is a valid option if your files don't exceed the overhead of a regular HTML response. AND/OR don't happen that often.
Don't get me wrong. I'm with you, generally this is a bad idea. But it has also its use-cases.


> @spapas: yes maybe it does I do agree but you must be really careful and know that your site may break

> @allcaps: Yes, as soon as someone can request a big enough file a couple of times and you occupy all your Python processes, you are in big trouble. So this option needs a big fat disclaimer.

> @spapas: we have a private section on our wagtail site but we only serve private docs not images. if it was supported we'd probably use it though (right now any images on the private pages are either served as documents or go t othe public place)

> @allcaps: The Wagtail related use-case is probably an intranet, where you mark a page for logged-in users only, but like to keep the images and documents also private.
All images and documents are behind access control. And images and documents inside the intranet collection are only served to authenticated users.

> @thibaudcolas: Is there any reason why Wagtail’s built-in collection privacy settings only apply to documents? I’m surprised we’d have done that by choice, is it just that images are more complex due to renditions? :thinking_face:

> @spapas: i'm not really sure ; I know that documents have a configurable "serve" method (which you can use to apply security checks) while images don't

> @spapas: maybe because of extra complexity maybe because it wasn't needed in a specific use case

> @allcaps: A Rendition has an url method, which allows you to point it to anywhere.

```python
class CustomRendition(AbstractRendition):
    # Override file, use private storage.
    file = models.ImageField(
        upload_to=get_rendition_upload_to,
        storage=get_private_storage,
        width_field="width",
        height_field="height",
        max_length=255,
    )
    image = models.ForeignKey(
        CustomImage,
        on_delete=models.CASCADE,
        related_name="renditions",
        verbose_name=_("image"),
    )

    class Meta:
        unique_together = (("image", "filter_spec", "focal_point_key"),)

    @property
    def url(self):
        value = super(CustomRendition, self).url.replace("/images/", "")
        return f"/images/{self.pk}/{value}"
```

> @spapas: https://docs.wagtail.org/en/stable/reference/settings.html#wagtaildocs-serve-method
hmmm so this could be used to serve the file through django-sendfile2 ?


> @allcaps: Yes. These are the (slightly complex because of requirements) views:

```python
from pathlib import Path

from django.http import Http404, HttpResponse
from django.shortcuts import get_object_or_404
from django_sendfile import sendfile
from wagtail import hooks
from wagtail.documents import get_document_model

from cwi.apps.custom_media.models import CustomRendition


def document_serve(request, pk, filename):
    """
    document_serve, a re-implementation of `wagtail.documents.views.serve.serve`

    Diff to `wagtail.documents.views.serve.serve`

    - No cache control
    - Works with sendfile2 (sendfile is not maintained)
    - Skips logic we don't use:
        - No document served signal
        - No `serve_method` lookup
        - No `direct_url` (remote data store)
        - No StreamingResponse
    """
    model = get_document_model()
    document = get_object_or_404(model, pk=pk)

    # We want to ensure that the document filename provided in the URL matches the
    # one associated with the considered document_id. If not we can't be sure that
    # the document the user wants to access is the one corresponding to the
    # <pk, filename> pair.
    if document.filename != filename:
        raise Http404("This document does not match the given filename.")

    # The permission check is handled by `before_serve_document` hooks.
    # See: `wagtail.documents.wagtail_hooks.check_view_restrictions`
    for fn in hooks.get_hooks("before_serve_document"):
        result = fn(document, request)
        if isinstance(result, HttpResponse):
            return result

    opts = {
        "attachment": (document.content_disposition != "inline"),
        "attachment_filename": document.filename,
        "mimetype": document.content_type,
    }

    # Sendfile2 needs a `local_path` as Path (not string).
    local_path = Path(document.file.path).resolve()
    return sendfile(request, local_path, **opts)


def image_serve(request, pk, filename):
    """
    image_serve, custom view to handle access control on renditions (images)

    - Inspired on `document_serve`
    - `before_serve_document` checks the Collection permissions
    """

    rendition = get_object_or_404(CustomRendition, pk=pk)

    # We want to ensure that the filename provided in the URL matches the
    # one associated with the considered rendition. If not we can't be sure that
    # the rendition the user wants to access is the one corresponding to the
    # <pk, filename> pair.
    if rendition.file.name.replace("images/", "") != filename:
        raise Http404("This rendition does not match the given filename.")

    # The permission check is handled by `before_serve_document` hooks.
    # See: `wagtail.documents.wagtail_hooks.check_view_restrictions`
    for fn in hooks.get_hooks("before_serve_document"):
        result = fn(rendition.image, request)
        if isinstance(result, HttpResponse):
            return result

    opts = {
        "attachment_filename": rendition.file.file,
    }
    local_path = Path(rendition.file.path).resolve()
    return sendfile(request, local_path, **opts)
```

> @spapas: very useful I didn't know that ! i'll probably use it (sometime) to add image support to our internal site :slightly_smiling_face:

> @thibaudcolas: There’s actually an open issue about support for images: https://github.com/wagtail/wagtail/issues/3880

> @allcaps: I forgot about this one :smile:
`before_serve_document` checks the Collection permissions
So, that permission check is there out-of-the-box.
That all media lives in a public directory `/media/` is the gotcha. It is logical that most servers are configured to serve all static and all media by default. But for any of this to work, your media should NOT live in a public dir.

**Reply by allcaps** [reply]
Ref https://docs.wagtail.org/en/stable/advanced_topics/deploying.html#user-uploaded-files
